[PyTorch] Reduce errors of `foreach` functions

crcrpar opened the pull request:
This is based on  #48224. 

To make `foreach` more flexible, this PR pushes unsupported cases to slow path.
Also, this adds some tests to verify that
- `foreach` functions work with tensors of different dtypes and/or memory layouts in https://github.com/pytorch/pytorch/commit/7bd4b2c89fad23c17a58969623ea7145833548a1
- `foreach` functions work with tensors on different devices in a list, but are on the same device if the indices are the same: https://github.com/pytorch/pytorch/commit/def4b9b5a19c325bb7f82ef6d69ca28fa2927131

Future plans:
1. Improve the coverage of unittests using `ops` decorator & updating `foreach_unary_op_db` and creating `foreach_(binary|pointwise|minmax)_db`.
2. Support broadcasting in slow path. Ref:  https://github.com/pytorch/pytorch/pull/52448
3. Support type promotion in fast path. Ref https://github.com/pytorch/pytorch/pull/52449

CC: @ngimel @mcarilli  @ptrblck

crcrpar committed:
Reduce errors

Ref: https://github.com/pytorch/pytorch/pull/48224

crcrpar committed:
update test upon foreach update

Apply skipMeta to reduce warnings

crcrpar committed:
more non-contiguous cases

crcrpar committed:
Test list of tensors on different devices

crcrpar committed:
make `check_dtype` optional in foreach api risrictions

crcrpar committed:
early retun of int/bool check

crcrpar committed:
foreach max/min-imum: cuda bool tensors on slowpath

crcrpar committed:
cosmetic

crcrpar committed:
fix abs invocation

crcrpar committed:
make tests of different device tensors simpler

ngimel commented on aten/src/ATen/native/cuda/ForeachPointwiseOp.cu:
```diff
@@ -105,7 +105,7 @@ std::vector<Tensor> foreach_pointwise_op(TensorList input, TensorList tensors1,
     check_foreach_api_restrictions(input, tensors1, tensors2);                                                                        \
                                                                                                                                       \
-    if (!can_use_fast_route({input, tensors1, tensors2}, scalar)) {                                                                   \
+    if (!can_use_fast_route({input, tensors1, tensors2}, scalar) || has_int_or_bool_tensor(input)) {                                  \
```
out of curiosity, why are we getting rid of int tensors? Because of type promotion?

crcrpar replied:
Yes, type promotion.

ngimel commented on aten/src/ATen/native/cuda/ForeachUnaryOp.cu:
```diff
@@ -247,14 +247,15 @@ struct Abs {
+    for (const auto& t : tensors) {
         if (at::isComplexType(t.scalar_type())) {
-            has_complex = true;
+            has_complex_or_integer = true;
```
this is misleading - how did `integer` get in the name? Also, std::any_of.

crcrpar replied:
ah, sorry, forgot to fix the name

crcrpar committed:
change name to has_complex

also, use `std::any_of`

crcrpar committed:
remove dtype checks from foreach api restrictions

Stop providing a special path for AMP kernels.

crcrpar committed:
fallback to single tensor kernel if dtypes differ

Upon the removal of dtype checks from `check_foreach_api_restrictions`,
non_finite_check_and_unscale_cuda_ internally checks each tensor's dtype
and push it to single-tensor TensorIterator kernel.

Also, update the comment about `can_use_fast_route`.

crcrpar committed:
update tests

_amp_foreach_non_finite_checks_and_unscale_ is expected to
- handle tensors of various dtypes
- raise a RuntimeError if devices are different

crcrpar commented:
check_foreach_api_restriction functions dropped the option of dtype checks and `_amp_foreach_non_finite_check_and_unscale_cuda_` is now able to handle tensorlists of mixed dtypes (this case is not supposed to happen though(.

ngimel commented on test/test_foreach.py:
```diff
@@ -806,6 +835,115 @@ def test_add_list_slow_path(self, device, dtype):
+            tensors1, tensors2, tensors3 = [[_cuda_tensors[i], _cpu_tensors[i]] for i in range(3)]
+
+            if native_op == torch.addcdiv:
+                if dtype in torch.testing.get_all_int_dtypes() + [torch.bool]:
```
can you please leave a comment here why this special case is necessary?

crcrpar replied:
Okay, I think the reason is similar to that of `sub and bool` case: the message has one or more special characters

ngimel reviewed (commented):
This looks good, I've left minor comments.

crcrpar commented on test/test_foreach.py:
```diff
@@ -126,6 +126,7 @@ def _test_bin_op_list_alpha(self, device, dtype, foreach_op, foreach_op_, torch_
             else:
                 self.assertEqual(tensors1, expected)
 
+    @skipMeta
```
I don't remember the details, but without this I saw a bunch of messages which seem like a ton of failures, but they were somehow treated as warnings or something like that.
The reason I added this decorator is, those messages were just saying "hey this operation is not working with meta tensor" but coming with the bunch of redundant information (at least for foreach tests).

ngimel replied:
It that's the case (the tests are actually passing), don't skip them please, so that if coverage is extended to meta they don't have to be explicitly reenabled. For you local runs you could still filter to only CPU or only CUDA device (not both together I'm afraid)

ezyang replied:
Yes, please don't do this. If you want to reduce the spamminess, edit the code in torch/testing/_internal/common_utils.py

```
@contextmanager
def skip_exception_type(exc_type):
    try:
        yield
    except exc_type as e:
        raise unittest.SkipTest(f"not implemented: {e}") from e
```

you could make it stop printing the entire exception text (though I've personally found it useful for collecting information from XML logs)

crcrpar replied:
Sure, thank you @ezyang for the explanation

crcrpar commented on test/test_foreach.py:
```diff
@@ -788,8 +809,16 @@ def test_add_list_different_sizes(self, device, dtype):
+    @skipMeta
     @dtypes(*torch.testing.get_all_dtypes())
     def test_add_list_slow_path(self, device, dtype):
+        # 0-strides
```
Sorry, just I'm wrong, but is it still the case, even if I deliberately call `expand_as(tensor1)`?
If sizes don't match, the current foreach functions should raise.

crcrpar replied:
Side note: I want to support broadcasting after test refactoring by pushing them to the slowpath.

Rel: https://github.com/pytorch/pytorch/pull/52448

ngimel replied:
If you deliberately expand, then it's explicit broadcast, but either way, it's just a comment, so doesn't matter

crcrpar commented on test/test_foreach.py:
```diff
@@ -806,6 +835,115 @@ def test_add_list_slow_path(self, device, dtype):
+            _cpu_tensors = self._get_test_data('cpu', dtype, 2)
+            tensors1, tensors2 = [[_cuda_tensors[i], _cpu_tensors[i]] for i in range(2)]
+
+            if dtype == torch.bool and native_op == torch.sub:
```
You are absolutely right and that is what I expected but it unfortunately fails.
If I merge this special handling into the following `try-except` block, in my understanding, as the error message includes the `^`, one of the regex special characters, I get the following:

```console
Traceback (most recent call last):
  File "/workspace/test/test_foreach.py", line 890, in test_binary_op_tensors_on_different_devices
    actual = foreach_op(tensors1, tensors2)
RuntimeError: Subtraction, the `-` operator, with two bool tensors is not supported. Use the `^` or `logical_xor()` operator instead.

During handling of the above exception, another exception occurred:

RuntimeError: Subtraction, the `-` operator, with two bool tensors is not supported. Use the `^` or `logical_xor()` operator instead.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/workspace/test/test_foreach.py", line 893, in test_binary_op_tensors_on_different_devices
    [native_op(t1, t2) for t1, t2 in zip(tensors1, tensors2)]
  File "/opt/conda/lib/python3.8/unittest/case.py", line 240, in __exit__
    self._raiseFailure('"{}" does not match "{}"'.format(
  File "/opt/conda/lib/python3.8/unittest/case.py", line 164, in _raiseFailure
    raise self.test_case.failureException(msg)
AssertionError: "Subtraction, the `-` operator, with two bool tensors is not supported. Use the `^` or `logical_xor()` operator instead." does not match "Subtraction, the `-` operator, with two bool tensors is not supported. Use the `^` or `logical_xor()` operator instead."
```

ngimel replied:
Can you try if wrapping in `re.escape` solves this? e.g.
```
with self.assertRaisesRegex(type(e), re.escape(str(e))):
```
re.search seems to be happy:
```In [10]: aa="pytorch `^`"

In [11]: re.search(re.escape(aa), bb)
Out[11]: <re.Match object; span=(0, 11), match='pytorch `^`'>

In [12]: re.search(aa, bb)

```

crcrpar replied:
It works!

crcrpar commented on aten/src/ATen/native/cuda/ForeachPointwiseOp.cu:
```diff
@@ -149,10 +149,20 @@ FOREACH_POINTWISE_OP_SCALAR(addcdiv, std::divides);
 FOREACH_POINTWISE_OP_SCALARLIST(addcdiv, std::divides);
 
+
+bool has_bool_tensor(TensorList tensors) {
```
That makes sense, I'll rename `has_int_or_bool_tensor` to `has_integral_tensor` and add `includeBool` argument.
I also will juxtapose `has_bool_tensor` with `has_integral_tensor`, but I want to have them as separate functions as the caller of `has_bool_tensor` can handle int tensors.
Or, is it better to write `std::any_of(tensor.begin(), tensors.end(), [](const auto & t) -> bool { t.scalar_type() == Bool;});` inside the caller?

crcrpar commented on test/test_foreach.py:
```diff
@@ -806,6 +835,115 @@ def test_add_list_slow_path(self, device, dtype):
+            # `tensors2`: ['cuda', 'cpu']
+            _cuda_tensors = self._get_test_data(device, dtype, 2)
+            _cpu_tensors = self._get_test_data('cpu', dtype, 2)
+            tensors1, tensors2 = [[_cuda_tensors[i], _cpu_tensors[i]] for i in range(2)]
```
Oh, I like this, thank you.

crcrpar committed:
Add comments & use Pythonic inputs creation

crcrpar committed:
juxtapose has_bool_tensor with has_integral_tensor

note: `has_int_or_bool_tensor` is renamed to `has_integral_tensor`
and a new argument `includeBool` is added.

crcrpar committed:
explain why bool tensors are pushed to slowpath

crcrpar committed:
comments and `will_promote_tensor` refactor

crcrpar committed:
comment about the limitation binary ops have

Binary foreach functions cannot be compiled for the inputs of
one tensorlist and one scalarlist if scalarlist consists of complex128.

crcrpar committed:
remove `skipMeta` decorator

leave a comment to the test which requires the decorator

crcrpar committed:
use `re.escape`

This obviates the special case for `_foreach_sub` and `_foreach_addcdiv`

ngimel reviewed (approved):
Looks good!
